File: clients/windows/Mailcal.Tests/AllodiaSubscriptionFixture.cs
// The account-service answers the subscription card reads, built here so each test names only the
// part it is about.
//
// Everything in `purchasing.md` that a card can get wrong is a shape of this record: who is
// billing, whether anything renews, whether a charge is being retried, and what the service says
// the caller may do. A fixture keeps those legible instead of burying each rule under nine fields.

using uniffi.mailcal_bindings;

namespace Allodia.Mailcal.Tests;

internal static class AllodiaSubscriptionFixture
{
    internal const string PeriodEnd = "2026-10-11T00:00:00+02:00";

    /// <summary>Nobody may do anything, which is what the service says about a free account.</summary>
    internal static AllodiaSubscriptionActions NoActions() => new(false, false, false, false);

    internal static AllodiaPrices Prices() => new(499, 4999, "EUR");

    /// <summary>A subscription Allodia bills directly, renewing, with nothing else charging.</summary>
    internal static AllodiaSubscription Own(
        AllodiaOwnStatus? status = null,
        AllodiaPlan interval = AllodiaPlan.Monthly,
        string? nextPayment = PeriodEnd,
        AllodiaSubscriptionActions? actions = null) =>
        new(
            Entitled: true,
            Plan: "standard",
            CurrentPeriodEnd: PeriodEnd,
            Own: new AllodiaOwnSubscription(
                Status: status ?? new AllodiaOwnStatus.Active(),
                Interval: interval,
                AmountInCents: 499,
                NextPaymentDate: nextPayment,
                CurrentPeriodEnd: PeriodEnd,
                CancelledAt: null),
            Stores: [],
            DuplicateBilling: [],
            Actions: actions ?? new AllodiaSubscriptionActions(true, false, true, false),
            Prices: Prices(),
            CheckoutAvailable: true);

    /// <summary>A subscription a store is billing, which this API may only read.</summary>
    internal static AllodiaStoreSubscription Store(
        AllodiaStore source,
        AllodiaStoreStatus? status = null,
        bool autoRenewing = true) =>
        new(
            Source: source,
            Status: status ?? new AllodiaStoreStatus.Active(),
            Interval: AllodiaPlan.Monthly,
            ProductId: "eu.allodia.mailcal.monthly",
            PriceInCents: 599,
            Currency: "EUR",
            CurrentPeriodEnd: PeriodEnd,
            AutoRenewing: autoRenewing,
            Environment: "production",
            ManageUrl: "https://example.test/manage");

    /// <summary>An account a store is billing, and nothing of Allodia's own.</summary>
    internal static AllodiaSubscription Stores(
        params AllodiaStoreSubscription[] stores) =>
        new(
            Entitled: true,
            Plan: "standard",
            CurrentPeriodEnd: PeriodEnd,
            Own: null,
            Stores: stores,
            DuplicateBilling: [],
            // Everything false: a store's subscription is changed at that store, and the service is
            // what says so.
            Actions: NoActions(),
            Prices: Prices(),
            CheckoutAvailable: true);

    /// <summary>Nothing is being charged, so both periods are on offer.</summary>
    internal static AllodiaSubscription Free() =>
        new(
            Entitled: false,
            Plan: "free",
            CurrentPeriodEnd: null,
            Own: null,
            Stores: [],
            DuplicateBilling: [],
            Actions: new AllodiaSubscriptionActions(false, false, false, true),
            Prices: Prices(),
            CheckoutAvailable: true);
}

File: clients/android/app/src/test/java/eu/allodia/mailcal/AllodiaSubscriptionFixture.kt
// One subscription answer to build the subscription tests out of, so each test states only the
// part it is about.
package eu.allodia.mailcal

import uniffi.mailcal_bindings.AllodiaBiller
import uniffi.mailcal_bindings.AllodiaOwnStatus
import uniffi.mailcal_bindings.AllodiaOwnSubscription
import uniffi.mailcal_bindings.AllodiaPlan
import uniffi.mailcal_bindings.AllodiaPrices
import uniffi.mailcal_bindings.AllodiaStore
import uniffi.mailcal_bindings.AllodiaStoreStatus
import uniffi.mailcal_bindings.AllodiaStoreSubscription
import uniffi.mailcal_bindings.AllodiaSubscription
import uniffi.mailcal_bindings.AllodiaSubscriptionActions

internal fun subscription(
    entitled: Boolean = false,
    currentPeriodEnd: String? = null,
    own: AllodiaOwnSubscription? = null,
    stores: List<AllodiaStoreSubscription> = emptyList(),
    duplicateBilling: List<AllodiaBiller> = emptyList(),
) = AllodiaSubscription(
    entitled = entitled,
    plan = if (entitled) "pro" else "free",
    currentPeriodEnd = currentPeriodEnd,
    own = own,
    stores = stores,
    duplicateBilling = duplicateBilling,
    // Permissive by default: what a client may offer is the service's answer, so a test about
    // anything else should not have to state it, and the tests that are about it say so.
    actions = AllodiaSubscriptionActions(
        canCancel = true,
        canResubscribe = true,
        canSwitchInterval = true,
        canStartCheckout = !entitled,
    ),
    prices = AllodiaPrices(monthlyInCents = 299, yearlyInCents = 2990, currency = "EUR"),
    checkoutAvailable = true,
)

internal fun storeSubscription(
    source: AllodiaStore = AllodiaStore.GOOGLE,
    status: AllodiaStoreStatus = AllodiaStoreStatus.Active,
    autoRenewing: Boolean = true,
    currentPeriodEnd: String? = null,
    manageUrl: String = "https://play.google.com/store/account/subscriptions",
) = AllodiaStoreSubscription(
    source = source,
    status = status,
    interval = AllodiaPlan.YEARLY,
    productId = "eu.allodia.mailcal.services",
    priceInCents = null,
    currency = null,
    currentPeriodEnd = currentPeriodEnd,
    autoRenewing = autoRenewing,
    environment = "production",
    manageUrl = manageUrl,
)

internal fun ownSubscription(
    status: AllodiaOwnStatus = AllodiaOwnStatus.Active,
    nextPaymentDate: String? = "2027-09-18T00:00:00+02:00",
    interval: AllodiaPlan = AllodiaPlan.YEARLY,
) = AllodiaOwnSubscription(
    status = status,
    interval = interval,
    amountInCents = 2990,
    nextPaymentDate = nextPaymentDate,
    currentPeriodEnd = "2027-09-18T00:00:00+02:00",
    cancelledAt = null,
)

// A service that permits none of the writes, which is what it answers for somebody a store is
// charging: they have a period and a plan, and neither is this API's to change.
internal fun refusing() = AllodiaSubscriptionActions(
    canCancel = false,
    canResubscribe = false,
    canSwitchInterval = false,
    canStartCheckout = false,
)
